src/basic/macro.h

```diff
@@ -449,4 +449,61 @@ assert_cc(sizeof(dummy_t) == 0);
                 _b_;                                                    \
         })
 
+/* Bit index(es) (0-based) to mask (of specified type). Assertion failure if (any) index is out of range. */
+#define INDEXES_TO_MASK(type, ...) _INDEXES_TO_MASK(type, UNIQ_T(_mask_, UNIQ), UNIQ_T(_indexes_, UNIQ), UNIQ_T(_i_, UNIQ), ##__VA_ARGS__)
+#define __INDEXES_TO_MASK(type, _mask_, _indexes_, _i_, ...)             \
+        ({                                                              \
+                type _mask_ = 0;                                        \
+                size_t _indexes_[] = { __VA_ARGS__ };                   \
+                for (size_t _i_ = 0; _i_ < ELEMENTSOF(_indexes_); _i_++) \
+                        _mask_ |= INDEX_TO_MASK(type, _indexes_[_i_];   \
+                _mask_;                                                 \
+        })
+#define _INDEXES_TO_MASK(type, _mask_, _indexes_, _i_, ...)             \
+        ({                                                              \
+                type _mask_ = 0;                                        \
+                VA_ARGS_FOREACH(_i_, size_t, __VA_ARGS__)               \
+                        _mask_ |= INDEX_TO_MASK(type, _i_);             \
+                _mask_;                                                 \
+        })
+#define INDEX_TO_MASK(type, index) _INDEX_TO_MASK(type, index, UNIQ_T(_index_, UNIQ))
+#define _INDEX_TO_MASK(type, index, _index_)                            \
+        ({                                                              \
+                size_t _index_ = (index);                               \
+                assert(sizeof(type) <= sizeof(unsigned long long));     \
+                assert(_index_ < sizeof(type) * 8);                     \
+                ((type)1) << _index_;                                   \
+        })
```

Comment:
definitions like this always have the problem they don't work in constant contexts, and for example when used in array sizing turn fixed size arrays into VLAs without people noticing. They might still be ok, but we should think twice.

Reply:
sorry im not following, what do you mean?

Reply:
to give an example:

```c
#define FOO(x) ({ x << 1; })
#define BAR(x) ((x) << 1) 

int main(int argc, char *argv[] {
        int a[FOO(7)];
        int b[BAR(7)];
        static int c[FOO(7)];    /* ← compile error */
        static int d[BAR(7)];
}
```

superficially, FOO() and BAR() in the code above do the same, i.e. return the input shifted left by one. But VAR() is cannot be evaluated in constant context. This means: b[] is going to be a regular array on the stack, but a[] is going to be a C99 variable length array (VLA) which everyone hates. d[] is going to be a regular static aray, but c[] is going to result in a compiler error since you cannot have static VLAs.

now the bad thing really is the VLA stuff: without the user realizing that they defined a VLA even though the size is of course conceptually totally constant.

hence what I am saying: for stuff that "feels" constant we really should try to return constant values, and `({ })` always destroys constants hence should be used with care.

Reply:
ok i see what you are saying; i'll see if i can make this constant-foldable